contentcuration/contentcuration/frontend/shared/app.js

```diff
@@ -122,5 +123,16 @@ export default async function startApp({ store, router, index }) {
     Object.assign(config, Base);
   }
 
+  window.addEventListener('beforeunload', e => {
+    const logoutConfirmed = window.sessionStorage.getItem('logoutConfirmed');
+    const areAllChangesSaved = store.getters['areAllChangesSaved'];
+
+    if (!logoutConfirmed && !areAllChangesSaved) {
+      forceServerSync();
```

Comment:
You mentioned that every call here has to be synchronous, so are we basically preventing navigation here in the case where there are unsynced changes? Presumably this means that user will have to renavigate to actually leave the page?

Reply:
I will answer this in more detail later. To briefly clarify:

> You mentioned that every call here has to be synchronous

When saying this:

> One of those limitations is that decision logic related to preventing unload needs to be synchronous.

I didn't mean that every call needs to be synchronous. It's possible to run async calls. But it's not possible to wait for its results for purpose of making decisions on whether to prevent the unload event or not - in code that means that result of async call can't be used in the condition on line 130. That's why those conditions and logic around them are build in a synchronous way.

Reply:
Right, my question was about the `e.preventDefault();`  call on the line below - this presumably means that the navigation is cancelled in this case, yes? Because if we didn't cancel the navigation, we would make the async call, but then we would navigate away from the page before it completed, so the request would not complete.

Reply:
> this presumably means that the navigation is cancelled in this case, yes?

In a sense, yes, though to be more precise this will rather cause the "Unsaved changes" prompt to get displayed. Then it's up to a user if they really want to cancel the navigation or not. In either of those two cases, server sync is called. If they decide to leave even though there were unsaved changes, who knows, maybe we are lucky and the sync finishes meanwhile anyways. If they decide to stay, it will finish (and if it's super slow and they try to leave again very fast, the prompt will be displayed again).

Reply:
OK, in that case, I would prefer to defer the sync call until they have clicked 'cancel' for the navigation in order to prevent the request getting cancelled and having an uncertain state regarding the unsynced changes (potentially some will have been committed to the server).

Reply:
We should invoke `window.confirm` here also, in the case that a logout hasn't been confirmed, warn about unsaved changes, and then either let them navigate, or cancel and initiate a sync.

Reply:
> We should invoke window.confirm here also, in the case that a logout hasn't been confirmed, warn about unsaved changes, and then either let them navigate, or cancel and initiate a sync.

I don't follow  what do you mean "in the case that a logout hasn't been confirmed" - could you describe it from the user point of view? What behavior are you experiencing?

Displaying custom prompts in [`beforeunload` doesn't usually work](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event):

> The HTML specification states that calls to window.alert(), window.confirm(), and window.prompt() methods may be ignored during this event. See the HTML specification for more details.

The default browser prompt is displayed when `beforeunload` event is prevented by calling `preventDefault`.

Reply:
I tried to describe the differences between leaving by logout and leaving by closing the tab in the implementation notes. Does it address this concern or are you referencing something else?

Reply:
> I would prefer to defer the sync call until they have clicked 'cancel' for the navigation in order to prevent the request getting cancelled and having an uncertain state regarding the unsynced changes

Your concern regarding an uncertain state makes sense.

I am not aware of any way of tracking if a user clicked "Cancel" when a default browser prompt is displayed after `beforeunload` is prevented and haven't been able to find a solution yet.

Though I think that I could just simply remove `forceServerSync` call [from here](https://github.com/learningequality/studio/pull/2699/files/26bded01ffe0ae817a55fb72c468b4683d1c1328#diff-f5a688eac83b9eaeb8320505c4528603197241c5da09b92d916c5d0ff0724d9cR131). If I can remember well, I added it only because I thought that it might be a little improvement and save some time. But  our sync runs every 2 seconds so it should happen soon anyway.

Reply:
Yeah - I think not trying to initiate server sync during the beforeunload handler seems sensible, I think that addresses my main concern of the state uncertainty.